### ISSUE TITLE
Support output to stdout for improved pipe support {feature request}

### ISSUE THREAD

**Issue by grill05** [OpenVisualCloud/SVT-AV1#701]
Currently, when using ffmpeg with SvtAv1EncApp with pipes, one does

`ffmpeg -i input.mkv -an -s 424x240 -f yuv4mpegpipe -pix_fmt yuv420p  - | SvtAv1EncApp -i stdin -n 10 -w 424 -h 240  -fps 24 -rc 3 -tbr 75000 -enc-mode 5  test.webm`

for producing output to a file test.webm
The problem here is that one cannot encode audio simultaneously. If Svt could output to stdout, one could do 

`ffmpeg -i input.mkv -an -s 424x240 -f yuv4mpegpipe -pix_fmt yuv420p  - | SvtAv1EncApp -i stdin -n 10 -w 424 -h 240  -fps 24 -rc 3 -tbr 75000 -enc-mode 5  - | ffmpeg -i - -i input.mkv -map 0:v -c:v copy -map 1:a -c:a libopus output.webm`

and in this case, output.webm would have both audio and video.
I know that there is a ffmpeg plugin (libsvt), but I cant get that to compile, so I have to use SvtAv1EncApp via pipes. 

It would be nice if svt could output to stdout (like rav1e and aomenc can)

**Comment by 1480c1** [comment]
Related to https://github.com/OpenVisualCloud/SVT-AV1/issues/302

Duplicate of https://github.com/OpenVisualCloud/SVT-AV1/issues/231

**Comment by EwoutH** [comment]
@hassount Could we get this in for the [0.7.5 release](https://github.com/OpenVisualCloud/SVT-AV1/milestone/1)? Would be a great addition to enlarge the user base (and thus testing and bug reports).

Also a useful step for #259.

**Comment by hassount** [comment]
@EwoutH getting it in 0.7.5 is probably not going to be possible as we won't have resources to support this feature. Though, this is high level enough to be open for contributions.

**Comment by j-mas** [comment]
@hassount@EwoutH
I think this is the most important feature for us users in the nearest future!

**Comment by grill05** [comment]
From a user perspective, I'd say 2-pass vbr mode is absolutely essential, while crf-mode and output to stdout would be very nice to have. 
That is because the encoder output looks very bad (visually) for high motion scenes right now (at low-ish bit rates). Whenever there is high motion/scene-change the encoder produces very visible pixellation (at low bit rates). Hopefully 2-pass vbr should help with that. 
Output to stdout lets one do audio+video encoding simultaneously (since the ffmpeg plugin gives segfaults).Of course, if the ffmpeg plugin is fixed, then one wouldn't need to use SvtAv1EncApp anyway.

**Comment by kylophone** [comment]
Implemented as follows. There is a larger problem, however. There is all sorts of nonsense being logged to stdout in the library and exe via `printf()` which will result in mangled bitstreams. So before we can actually have this feature, that needs to be cleaned up. There are two ways that can be dealt with, as outlined below, my preference being for that latter. This should be a relatively simple task. 

* use `fprintf(stderr, ...)` for all logging.
* implement a real logger, as suggested in #302

```
diff --git a/Source/App/EncApp/EbAppConfig.c b/Source/App/EncApp/EbAppConfig.c
index 963743e1..758b35d3 100644
--- a/Source/App/EncApp/EbAppConfig.c
+++ b/Source/App/EncApp/EbAppConfig.c
@@ -159,11 +159,17 @@ static void SetCfgInputFile(const char *filename, EbConfig *cfg)
     cfg->y4m_input = check_if_y4m(cfg);
 };
 
-static void SetCfgStreamFile                    (const char *value, EbConfig *cfg)
+static void SetCfgStreamFile(const char *filename, EbConfig *cfg)
 {
-    if (cfg->bitstream_file) { fclose(cfg->bitstream_file); }
-    FOPEN(cfg->bitstream_file,value, "wb");
+    if (cfg->bitstream_file)
+        fclose(cfg->bitstream_file);
+
+    if (!strcmp(filename, "stdout"))
+        cfg->bitstream_file = stdout;
+    else
+        FOPEN(cfg->bitstream_file, filename, "wb");
 };
+
 static void SetCfgErrorFile                     (const char *value, EbConfig *cfg)
 {
     if (cfg->error_log_file) { fclose(cfg->error_log_file); }
```

**Comment by xuguangxin** [comment]
Hi all
#302 addressed by https://github.com/OpenVisualCloud/SVT-AV1/pull/809. 

@kylophone , 
I used the following command to check "ffmpeg + libsvtav1 to stdout". It works very well.

> ffmpeg -i h264.mp4 -c:v libsvt_av1 -f ivf - |ffmpeg -i - out.yuv -y

After #809 merged, could you help rebase your patch on top of it?

thanks

**Comment by grill05** [comment]
@xuguangxin would this sort of pipe work with SvtAv1EncApp too ?

**Comment by xuguangxin** [comment]
Not yet, it needs this patch from @kylophone .
Could you share with us why you can't compile ffmpeg + libsvtav1?
any building issue?
thanks

> Implemented as follows. There is a larger problem, however. There is all sorts of nonsense being logged to stdout in the library and exe via `printf()` which will result in mangled bitstreams. So before we can actually have this feature, that needs to be cleaned up. There are two ways that can be dealt with, as outlined below, my preference being the latter. This should be a relatively simple task.
> 
> * use `fprintf(stderr, ...)` for all logging.
> * implement a real logger, as suggested in #302
> 
> ```diff
> diff --git a/Source/App/EncApp/EbAppConfig.c b/Source/App/EncApp/EbAppConfig.c
> index 963743e1..758b35d3 100644
> --- a/Source/App/EncApp/EbAppConfig.c
> +++ b/Source/App/EncApp/EbAppConfig.c
> @@ -159,11 +159,17 @@ static void SetCfgInputFile(const char *filename, EbConfig *cfg)
>      cfg->y4m_input = check_if_y4m(cfg);
>  };
>  
> -static void SetCfgStreamFile                    (const char *value, EbConfig *cfg)
> +static void SetCfgStreamFile(const char *filename, EbConfig *cfg)
>  {
> -    if (cfg->bitstream_file) { fclose(cfg->bitstream_file); }
> -    FOPEN(cfg->bitstream_file,value, "wb");
> +    if (cfg->bitstream_file)
> +        fclose(cfg->bitstream_file);
> +
> +    if (!strcmp(filename, "stdout"))
> +        cfg->bitstream_file = stdout;
> +    else
> +        FOPEN(cfg->bitstream_file, filename, "wb");
>  };
> +
>  static void SetCfgErrorFile                     (const char *value, EbConfig *cfg)
>  {
>      if (cfg->error_log_file) { fclose(cfg->error_log_file); }
> ```

**Comment by grill05** [comment]
@xuguangxin thanks for this feature. Last time I checked, enabling both libaom and libsvt-av1 produced segmentation faults in ffmpeg (this issue https://github.com/OpenVisualCloud/SVT-AV1/issues/567)
So right now, if I want to build ffmpeg with libaom support, then the only way to use SVT-AV1 is to pipe y4m to SvtAv1WEncApp

**Comment by xuguangxin** [comment]
could you double-check the #567 ? I can't reproduce it anymore https://github.com/OpenVisualCloud/SVT-AV1/issues/567#issuecomment-555358889

thanks

**Comment by grill05** [comment]
@xuguangxin nope. I still get a segfault. Test details under your comment in that particular issue

**Comment by xuguangxin** [comment]
@kylophone ,  https://github.com/OpenVisualCloud/SVT-AV1/pull/809 merged. could you help send pr for your stdout patch?
thanks
> Implemented as follows. There is a larger problem, however. There is all sorts of nonsense being logged to stdout in the library and exe via `printf()` which will result in mangled bitstreams. So before we can actually have this feature, that needs to be cleaned up. There are two ways that can be dealt with, as outlined below, my preference being the latter. This should be a relatively simple task.
> 
> * use `fprintf(stderr, ...)` for all logging.
> * implement a real logger, as suggested in #302
> 
> ```diff
> diff --git a/Source/App/EncApp/EbAppConfig.c b/Source/App/EncApp/EbAppConfig.c
> index 963743e1..758b35d3 100644
> --- a/Source/App/EncApp/EbAppConfig.c
> +++ b/Source/App/EncApp/EbAppConfig.c
> @@ -159,11 +159,17 @@ static void SetCfgInputFile(const char *filename, EbConfig *cfg)
>      cfg->y4m_input = check_if_y4m(cfg);
>  };
>  
> -static void SetCfgStreamFile                    (const char *value, EbConfig *cfg)
> +static void SetCfgStreamFile(const char *filename, EbConfig *cfg)
>  {
> -    if (cfg->bitstream_file) { fclose(cfg->bitstream_file); }
> -    FOPEN(cfg->bitstream_file,value, "wb");
> +    if (cfg->bitstream_file)
> +        fclose(cfg->bitstream_file);
> +
> +    if (!strcmp(filename, "stdout"))
> +        cfg->bitstream_file = stdout;
> +    else
> +        FOPEN(cfg->bitstream_file, filename, "wb");
>  };
> +
>  static void SetCfgErrorFile                     (const char *value, EbConfig *cfg)
>  {
>      if (cfg->error_log_file) { fclose(cfg->error_log_file); }
> ```

**Comment by sandersaares** [comment]
This feature is valuable to me, so I am interested in helping it move forward. Is there anything else that needs doing besides applying the above patch?
